sandbox/cgroup/cgroup.go

```diff
@@ -259,3 +259,110 @@ func MockVersion(mockVersion int, mockErr error) (restore func()) {
 		probeVersion, probeErr = oldVersion, oldErr
 	}
 }
+
+// procInfoEntry describes a single line of /proc/PID/cgroup.
+//
+// CgroupID is the internal kernel identifier of a mounted cgroup.
+// Controllers is a list of controllers in a specific cgroup
+// Path is relative to the cgroup mount point.
+//
+// Cgroup mount point is not provided here. It must be derived by
+// cross-checking with /proc/self/mountinfo. The identifier is not
+// useful for this.
+//
+// Cgroup v1 have non-empty Controllers and CgroupId > 0.
+// Cgroup v2 have empty Controllers and CgroupId == 0
+type procInfoEntry struct {
+	CgroupID    int
+	Controllers []string
+	Path        string
+}
+
+var pathOfProcPidCgroup = func(pid int) string {
+	return fmt.Sprintf("/proc/%d/cgroup", pid)
+}
+
+// MockPathOfProcPidCgroup mocks the function used to compute /proc/PID/cgroup
+func MockPathOfProcPidCgroup(fn func(int) string) func() {
+	old := pathOfProcPidCgroup
+	pathOfProcPidCgroup = fn
+	return func() {
+		pathOfProcPidCgroup = old
+	}
+}
+
+// ProcessPathInTrackingCgroup returns the path in the hierarchy of the tracking cgroup.
+//
+// Tracking cgroup is whichever cgroup systemd uses for tracking processes.
+// On modern systems this is the v2 cgroup. On older systems it is the
+// controller-less name=systemd cgroup.
+//
+// This function fails on systems where systemd is not used and subsequently
+// cgroups are not mounted.
+func ProcessPathInTrackingCgroup(pid int) (string, error) {
+	fname := pathOfProcPidCgroup(pid)
+	entries, err := loadProcCgroup(fname)
+	if err != nil {
+		return "", err
+	}
+	for _, e := range entries {
```

Comment:
We're only interested in id=0 or named controller, so maybe instead of reading the whole file, parsing and allocating a list of wrappers we could just look at the string directly?

```
for scanner.Scan() {
    group := scanner.Text()
    // 2 scenarios possible, 0::<path> on a cgroupv2 or hybrid, 
    // or <id>:name=systemd:<path> on a v1 or hybrid
    sysdidx := strings.Index(":name=systemd:")
    if !strings.HasPrefix("0::") && sysdidx == -1 {
        continue
    }
    if sysdidx != -1 {
        return line[sysdidx + len(":name=systemd:"):], nil
    }
    // 0:: prefix
    return line[len("0::"):], nil
}
```

Reply:
This is green, can I rework it on the next pass?

Reply:
Yeah, sounds good to me.